internal/controller/storageclient_controller.go

```diff
@@ -692,16 +692,33 @@ func (r *storageClientReconcile) reconcileResource(obj client.Object, rawObject
 		}
 		return nil
 	}
-
 	var err error
-	if useReplace {
-		err = utils.CreateOrReplace(r.ctx, r.Client, obj, mutateFunc)
-	} else {
-		_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, obj, mutateFunc)
-	}
+	_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, obj, mutateFunc)
+	if utils.IsForbiddenError(err) {
+		key := client.ObjectKeyFromObject(obj)
+		if err := r.Client.Get(r.ctx, key, obj); client.IgnoreNotFound(err) != nil {
+			return fmt.Errorf("failed to replace %v %v: %v", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), err)
+		}
+		if err := mutateFunc(); err != nil {
+			return fmt.Errorf("failed to replace %v %v: %v", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), err)
+		}
+		if newKey := client.ObjectKeyFromObject(obj); key != newKey {
+			return fmt.Errorf("failed to replace %v %v: MutateFn cannot mutate object name and/or object namespace", obj.GetObjectKind().GroupVersionKind(), obj.GetName())
+		}
 
-	if err != nil {
+		if err := r.Client.Delete(r.ctx, obj); client.IgnoreNotFound(err) != nil {
+			return fmt.Errorf("failed to replace %v %v: %v", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), err)
+		}
+
+		// k8s doesn't allow us to create objects when resourceVersion is set, as we are DeepCopying the
+		// object, the resource version also gets copied, hence we need to set it to empty before creating it
+		obj.SetResourceVersion("")
+		if err := r.Client.Create(r.ctx, obj); err != nil {
+			return fmt.Errorf("failed to replace %v %v: %v", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), err)
```

Comment:
You should not be doing delete and create inside CreateOrUpdate mutate func. CreateOrUpdate will do its own Update operation when the mutate func finishes

Reply:
The Create and delete is outside the CreateOrUpdate mutate function